src/vidpf.rs

```diff
@@ -0,0 +1,743 @@
+        let sl = conditional_xor_seeds(&seq_tilde.left_seed, s_cw, t_i);
+        let sr = conditional_xor_seeds(&seq_tilde.right_seed, s_cw, t_i);
+        let tl = seq_tilde.left_control_bit ^ (t_i & *t_cw_l);
+        let tr = seq_tilde.right_control_bit ^ (t_i & *t_cw_r);
+
+        let x_i = Choice::from(u8::from(
+            input.get_checked(level).ok_or(VidpfError::IndexLevel)?,
+        ));
+        let s_tilde_i = conditional_select_seed(x_i, &[sl, sr]);
+
+        let next_control_bit = Choice::conditional_select(&tl, &tr, x_i);
+        let (next_seed, w_i) = self.convert(s_tilde_i, nonce);
+
+        let zero = <W as IdpfValue>::zero(&self.weight_parameter);
+        let mut y = w_i + <W as IdpfValue>::conditional_select(&zero, w_cw, next_control_bit);
+        y.conditional_negate(Choice::from(id));
+
+        let pi_i = &state.proof;
+        let cs_i = &public.cs.get(level).ok_or(VidpfError::IndexLevel)?;
+        let pi_tilde = Self::node_proof(input, level, &next_seed)?;
+        let h2_input = pi_tilde.conditional_xor(next_control_bit, cs_i).xor(pi_i);
+        let next_proof = Self::node_proof_adjustment(h2_input).xor(pi_i);
+
+        let next_state = EvalState {
+            seed: next_seed,
+            control_bit: next_control_bit,
+            proof: next_proof,
+        };
+
+        Ok((next_state, y))
+    }
+
+    fn prg(seed: &Seed, nonce: &[u8]) -> PrgOutput {
+        let mut rng = XofFixedKeyAes128::seed_stream(&XofSeed(*seed), DomainSepTag::PRG, nonce);
+
+        let mut left_seed = Seed::default();
+        let mut right_seed = Seed::default();
+        rng.fill_bytes(&mut left_seed);
+        rng.fill_bytes(&mut right_seed);
+        // Use the LSB of seeds as control bits, and clears the bit,
+        // i.e., seeds produced by `prg` always have their LSB = 0.
+        // This ensures `prg` costs two AES calls only.
+        let left_control_bit = Choice::from(left_seed[0] & 0x1);
+        let right_control_bit = Choice::from(right_seed[0] & 0x1);
+        left_seed[0] &= 0xFE;
+        right_seed[0] &= 0xFE;
+
+        PrgOutput {
+            left_seed,
+            left_control_bit,
+            right_seed,
+            right_control_bit,
+        }
+    }
+
+    fn convert(&self, seed: Seed, nonce: &[u8; NONCE_SIZE]) -> (Seed, W) {
+        let mut rng = XofFixedKeyAes128::seed_stream(&XofSeed(seed), DomainSepTag::CONVERT, nonce);
+
+        let mut out_seed = Seed::default();
+        rng.fill_bytes(&mut out_seed);
+        let value = <W as IdpfValue>::generate(&mut rng, &self.weight_parameter);
+
+        (out_seed, value)
+    }
+
+    fn node_proof(input: &VidpfInput, level: usize, seed: &Seed) -> Result<Proof, VidpfError> {
+        let mut shake = XofTurboShake128::init(seed, DomainSepTag::NODE_PROOF);
+        for chunk128 in input
+            .index(..=level)
+            .chunks(128)
+            .map(BitField::load_le::<u128>)
+            .map(u128::to_le_bytes)
```

Comment:
We can make this slightly more compact and possibly save a Keccak evaluation
```suggestion
            .chunks(8)
            .map(BitField::load_le::<u8>)
```

Reply:
this is quite the opposite, we want to call SHAKE fewer times, not updating byte to byte. 
Also note that the blocksize of SHAKE is large (~168 bytes), so calls with 128 bits (16 bytes) is more efficient.

Reply:
I see. It's your call! FYI: we are likely to see input lengths of anywhere from 32 to 256 bits. We should also plan for 512 bits I think.